File: ruby/features/step_definitions/stepdefs.rb
require File.expand_path("lib/energy_market")

def load_plans(filepath)
    file = File.open(File.expand_path(filepath))
    data = JSON.load(file.read)
end

Given("the plans provided") do
    @market = EnergyMarket.new(load_plans("../plans.json"))
end

Given("plans with discounts") do
    @market = EnergyMarket.new(load_plans("../plans-with-discounts.json"))
end

When("annual usage is {int} kwh") do |usage|
    @prices = @market.price(usage)
end

Then("the cheapest plans are") do |table|
    table_hash = table.hashes
    
    expect(@prices[0]).to eq(table_hash[0])
    expect(@prices[1]).to eq(table_hash[1])
    expect(@prices[2]).to eq(table_hash[2])
    expect(@prices[3]).to eq(table_hash[3])
end

When("supplier name is {string}") do |string|
    @supplier_name = string
end
  
When("plan name is {string}") do |string|
    @plan_name = string
end
  
When("monthly spend is {int} pounds") do |int|
    @monthly_spend = int
end

When("monthly spend is {float} pounds") do |float|
    @monthly_spend = float
end
  
Then("annual usage is {int} kWh") do |int|
    usage = @market.usage(@supplier_name, @plan_name, @monthly_spend)
    expect(usage).to eq(int)
end

File: js/features/step_definitions/stepdefs.js
const fs = require('fs')
const assert = require('assert')

const { Given, When, Then } = require('cucumber')
const { price, usage } = require('../../lib/energy-market')

const loadPlans = function(filepath) {
  const json = fs.readFileSync(filepath, "utf8");
  return JSON.parse(json);
}

Given('the plans provided', function () {
  this.plans = loadPlans('../plans.json')
})

Given('plans with discounts', function () {
  this.plans = loadPlans('../plans-with-discounts.json')
})

/* Price */
When('annual usage is {int} kwh', function (usage) {
  this.usage = usage
})

Then('the cheapest plans are', function (plans) {
  const actual = price(this.plans, this.usage)
  const expected = plans.hashes()

  assert.deepStrictEqual(actual[0], expected[0])
  assert.deepStrictEqual(actual[1], expected[1])
  assert.deepStrictEqual(actual[2], expected[2])
  assert.deepStrictEqual(actual[3], expected[3])
})

/* Usage */
When('supplier name is {string}', function (supplierName) {
  this.supplierName = supplierName
})

When('plan name is {string}', function (planName) {
  this.planName = planName
})

When('monthly spend is {float} pounds', function (spend) {
  this.monthlySpend = spend
})

Then('annual usage is {int} kWh', function (expectedUsage) {
  const actualUsage = usage(
    this.plans,
    this.supplierName,
    this.planName,
    this.monthlySpend
  )

  assert.strictEqual(actualUsage, expectedUsage)
})
